### PR TITLE
feat: add support for testing S3 API locally 

### PR DESCRIPTION
**Pull request by akshbhu**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -17,4 +17,5 @@ packages/amplify-graphql*/lib
 packages/amplify-e2e-tests/lib
 packages/*/node_modules
 packages/*/package-lock.json
+packages/*/lib
```
Comment (yuth) [inline .gitignore]: Can you remove this. We have some JS packages which has some code in `lib` directory which is needed

File: packages/amplify-storage-simulator/package.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    "event-to-promise": "^0.8.0",
+    "express": "^4.17.1",
+    "glob": "^7.1.4",
+    "ip": "^1.1.5",
+    "lodash": "^4.17.14",
+    "md5-file": "^4.0.0",
+    "object-to-xml": "^2.0.0",
+    "save": "^2.4.0",
+    "serve-static": "^1.14.1",
+    "tsc-watch": "^2.2.1",
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: Do you need `tsc-watch`

File: packages/amplify-storage-simulator/package.json [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    "xml": "^1.0.1",
+    "xml-js": "^1.6.11"
+  },
+  "devDependencies": {
+    "@types/body-parser": "^1.17.0",
+    "@types/cors": "^2.8.5",
+    "@types/glob": "^7.1.1",
+    "@types/serve-static": "^1.13.2",
+    "typescript": "^3.5.3",
+    "typescript-logging": "^0.6.3"
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: Do you need `typescript-logging`

File: packages/amplify-storage-simulator/src/__tests__/S3server.test.ts [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+});
+
+describe("Test put api", () => {
+  const actual_file = __dirname + "/test-data/2.png";
+  const buffer = fs.readFileSync(actual_file);
+  test("put image", async () => {
+    const params = {
+      Bucket: bucket, // pass your bucket name
+      Key: "2.png",
+      Prefix: "upload", // file will be saved as testBucket/contacts.csv
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/__tests__/S3server.test.ts]: is the comment accurate?

File: packages/amplify-storage-simulator/src/__tests__/S3server.test.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    expect(data).toBeDefined();
+  });
+
+  const file = __dirname + "/test-data/abc.txt";
+  const buf1 = fs.readFileSync(file);
+
+  test("put text", async () => {
+    const params = {
+      Bucket: bucket, // pass your bucket name
+      Key: "upload/abc.txt", // file will be saved as testBucket/contacts.csv
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/__tests__/S3server.test.ts]: is the comment accurate?

File: packages/amplify-storage-simulator/src/__tests__/S3server.test.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    expect(data).toBeDefined();
+  });
+
+  const file1 = __dirname + "/test-data/Snake_River_(5mb).jpg";
+  const buf2 = fs.readFileSync(file1);
+
+  test(" multipart upload", async () => {
+    const params = {
+      Bucket: bucket, // pass your bucket name
+      Key: "upload/long_image.jpg", // file will be saved as testBucket/contacts.csv
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/__tests__/S3server.test.ts]: same here

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+      this.server.close();
+      this.server = null;
+      this.connection = null;
+    }
+  }
+
+  private async handleRequestAll(request, response) {
+    // parsing the path and the request parameters
+    request.url = decodeURIComponent(request.url);
+    //var str2 = this.route.slice(0, -1) + '';
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: remmove

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+    if (request.method === "PUT") {
+      this.handleRequestPut(request, response);
+    }
+
+    if (request.method === "POST") {
+      this.handleRequestPost(request, response);
+    }
+
+    if (request.method === "GET") {
+      if (request.params.path.indexOf(".") === -1) {
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts:118]: Can we do a check on if this request ends with `/` treat it as list other wise treat it as get. Also see if there exists a file with the key name treat it as get and if its a folder treat it as list

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+        o2x({
+          '?xml version="1.0" encoding="utf-8"?': null,
+          InitiateMultipartUploadResult: {
+            Bucket: this.route,
+            Key: request.params.path,
+            UploadId: this.uploadId
+          }
+        })
+      );
+    } else if (request.query.uploadId === this.uploadId) {
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts:269]: How do we handle concurrent uplodas from different clients

File: packages/amplify-util-mock/commands/mock/storage.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+const testUtil = require('../../lib');
+module.exports = {
+  name: 'storage',
+  run: async function(context) {
+    if (context.parameters.options.help) {
+      const header = `amplify test ${this.name} \nDescriptions:
```
Comment (yuth) [inline packages/amplify-util-mock/commands/mock/storage.js]: amplify mock ${this.name}

File: packages/amplify-util-mock/commands/mock/storage.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+      Test Storage locally`
```
Comment (yuth) [inline packages/amplify-util-mock/commands/mock/storage.js]: Mock storage locally

File: packages/amplify-util-mock/src/storage/storage.ts [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+      backendPath,
+      "storage",
+      resourceName,
+      "parameters.json"
+    );
+
+    const localEnvFilePath = context.amplify.pathManager.getLocalEnvFilePath();
+    const localEnvInfo = context.amplify.readJsonFile(localEnvFilePath);
+    const storageParams = context.amplify.readJsonFile(parametersFilePath);
+    this.bucketName = `${storageParams.bucketName}-${localEnvInfo.envName}`;
```
Comment (yuth) [inline packages/amplify-util-mock/src/storage/storage.ts:42]: Could we use the same location for storing files irrespective of what is the current env user is in. This is the current behavior for API

File: packages/amplify-velocity-template/src/parse/index.js [hunk 13]
```diff
@@ -71,2061 +71,439 @@
...
-      yytext,
-      yyleng,
-      yylineno,
-      yy,
-      yystate /* action[1] */,
-      $$ /* vstack */,
-      _$ /* lstack */
-    ) {
-      /* this == yyval */
+var index = (function(){
```
Comment (yuth) [inline packages/amplify-velocity-template/src/parse/index.js:74]: remove this file

File: .gitignore (hunk 1 above)
Comment (akshbhu) [reply .gitignore]: changed it to packages/amplify-storage-simulator/lib

File: packages/amplify-storage-simulator/package.json (hunk 2 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/package.json]: removed

File: packages/amplify-storage-simulator/package.json (hunk 3 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/package.json]: removed

File: packages/amplify-storage-simulator/src/__tests__/S3server.test.ts (hunk 4 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/__tests__/S3server.test.ts]: removed

File: packages/amplify-storage-simulator/src/__tests__/S3server.test.ts (hunk 5 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/__tests__/S3server.test.ts]: removed

File: packages/amplify-storage-simulator/src/__tests__/S3server.test.ts (hunk 6 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/__tests__/S3server.test.ts]: removed

File: packages/amplify-storage-simulator/src/server/S3server.ts (hunk 7 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: removed

File: packages/amplify-util-mock/commands/mock/storage.js (hunk 10 above)
Comment (akshbhu) [reply packages/amplify-util-mock/commands/mock/storage.js]: changed to mock

File: packages/amplify-util-mock/commands/mock/storage.js (hunk 11 above)
Comment (akshbhu) [reply packages/amplify-util-mock/commands/mock/storage.js]: changed to mock

File: packages/amplify-util-mock/src/storage/storage.ts (hunk 12 above)
Comment (akshbhu) [reply packages/amplify-util-mock/src/storage/storage.ts:42]: changed it to bucket name so that storing location remains same for diff envs

**Review by yuth — APPROVED** [review]

**Pull request merged**